Workaround needed for running Satellite in GitHub actions

## Context and discovering the issue
- After around 7 months of https://github.com/inform-us/INFORMus being dormant, we discovered a few weeks ago that our api tests that depend heavily on Satellite no longer run in GitHub actions (but they all still passed locally). [Our last successful CI run with all tests passing was in July 2025](https://github.com/inform-us/INFORMus/pull/2049). Unfortunately the logs for that CI run have now expired.
- For context, we build the Satellite image as part of our actions workflow (with actions running on `ubuntu-latest`) and spin it up with `docker compose` alongside our other services, then run pytest against it using the mapped port (5433:5432).
- The error seemed to come from a connection issue to the Satellite postgres instance. You can see a full log of the [CI failure in one of our PR's](https://github.com/inform-us/INFORMus/actions/runs/21594477161/job/62227833460). The strange thing was that this error persisted even though the `pg_isready` healthcheck passed:

```
Container informus-db  Healthy
Container informus-satellite  Healthy

...

E           psycopg2.OperationalError: server closed the connection unexpectedly
E           	This probably means the server terminated abnormally
E           	before or while processing the request.
```

## The fix
We added the following to your [recommended postgres setup](https://github.com/SAFEHR-data/Satellite/blob/main/README.md):

```diff                                                                                                    
   postgres:                                                                                                 
       image: satellite                                                                                      
  +    entrypoint: ["docker-entrypoint.sh"]                                                                  
  +    command: ["postgres"]                                                                                 
  +    restart: unless-stopped                                                                               
       container_name: informus-satellite                                                                    
       build:                                                                                                
         context: https://github.com/SAFEHR-data/Satellite.git                                               
   ...                                                                                                       
  ``` 

## Things we discovered
Our investigations might point towards an issue in the Postgres startup lifecycle.
- The `waveform_array` table creation has a syntax error (a trailing comma): `CREATE TABLE star.waveform_array (waveform_array_id serial PRIMARY KEY, );`. It is possible that this might be causing the init scripts to fail during the `docker-entrypoint.sh` setup phase. Perhaps this prevents Postgres from completing its startup correctly which causes the connection error? 
- One thing that came up a few times is that the Satellite container was refusing connections over TCP. So the custom docker commands we added do the following:
  - `entrypoint: ["docker-entrypoint.sh"]` and `command: ["postgres"]` bypasses the custom `entrypoint.sh` to keep TCP connections alive.
  - `restart: unless-stopped` recovers from errors (e.g. the syntax error due to `waveform_array`)
- It's possible that fixing the syntax error might remove the need for all three custom commands. But I'm not 100% sure as we don't have access to our historic logs and we don't know how long the syntax error has been there. If the syntax error existed before July 2025 then the root cause of the issue is likely something different.

Tagging @samcunliffe and @K-Meech who both helped with debugging this issue

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -33,7 +33,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout code
-        uses: actions/checkout@6
+        uses: actions/checkout@v6
         with:
           # Full git history is needed to get a proper list of
           # changed files within `super-linter`

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -49,7 +49,7 @@ jobs:
           extra_args: --all-files
 
       - name: Lint code base
-        uses: github/super-linter/slim@v8
+        uses: super-linter/super-linter/slim@v8
         env:
           VALIDATE_ALL_CODEBASE: false
           DEFAULT_BRANCH: main

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -49,7 +49,7 @@ jobs:
           extra_args: --all-files
 
       - name: Lint code base
-        uses: github/super-linter/slim@v8.4.0
+        uses: github/super-linter/slim@v8
         env:
           VALIDATE_ALL_CODEBASE: false
           DEFAULT_BRANCH: main

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -33,7 +33,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout code
-        uses: actions/checkout@63
+        uses: actions/checkout@6
         with:
           # Full git history is needed to get a proper list of
           # changed files within `super-linter`

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -33,7 +33,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout code
-        uses: actions/checkout@v3
+        uses: actions/checkout@63
         with:
           # Full git history is needed to get a proper list of
           # changed files within `super-linter`
@@ -41,15 +41,15 @@ jobs:
           persist-credentials: false
 
       - name: Setup Python
-        uses: actions/setup-python@v3
+        uses: actions/setup-python@v6
 
       - name: Run pre-commit
-        uses: pre-commit/action@v3.0.0
+        uses: pre-commit/action@v3.0.1
         with:
           extra_args: --all-files
 
       - name: Lint code base
-        uses: github/super-linter/slim@v4.9.7
+        uses: github/super-linter/slim@v8.4.0
         env:
           VALIDATE_ALL_CODEBASE: false
           DEFAULT_BRANCH: main

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -30,10 +30,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout code
-        uses: actions/checkout@v3
+        uses: actions/checkout@v6
 
       - name: Install Python
-        uses: actions/setup-python@v3
+        uses: actions/setup-python@v6
 
       - name: Install Python dependencies
         run: |

--- a/.gitignore
+++ b/.gitignore
@@ -6,3 +6,5 @@ satellite.egg-info/
 .mypy_cache/
 .pytest_cache/
 .DS_Store
+.vscode
+tmp*

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -16,19 +16,19 @@ fail_fast: true
 
 repos:
   - repo: https://github.com/ambv/black
-    rev: 22.12.0
+    rev: 26.1.0
     hooks:
       - id: black
         args: [--diff, --check]
 
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.991
+    rev: v1.19.1
     hooks:
       - id: mypy
         args: [--ignore-missing-imports]
 
   - repo: https://github.com/pycqa/flake8
-    rev: 6.1.0
+    rev: 7.3.0
     hooks:
       - id: flake8
 

--- a/Dockerfile
+++ b/Dockerfile
@@ -11,7 +11,7 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 # limitations under the License.
-FROM postgres:15.0-bullseye
+FROM postgres:17
 
 # User definable arguments
 ARG POSTGRES_USER=postgres
@@ -32,7 +32,7 @@ ARG DEBIAN_FRONTEND=noninteractive
 # OS setup
 RUN apt-get update && \
     apt-get install --yes --no-install-recommends \
-    procps ca-certificates locales python3.9-dev python3-pip git && \
+    procps ca-certificates locales pipx git && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/* && \
     sed -i '/en_GB.UTF-8/s/^# //g' /etc/locale.gen && locale-gen
@@ -46,11 +46,11 @@ RUN if [ -d "/Satellite/satellite" ] ; then \
       echo "Cloning Satellite repo on: $TAG" && \
       rm -rf /Satellite && \
       git clone --depth 1 --branch ${TAG} https://github.com/SAFEHR-data/Satellite.git ;\
-    fi && \
-    pip install --no-cache-dir --upgrade pip==22.3.1
+    fi
 
+ENV PATH="$PATH:/root/.local/bin"
 WORKDIR /Satellite
-RUN pip install --no-cache-dir . && \
+RUN pipx install . && \
     satellite print-db-create-command > /docker-entrypoint-initdb.d/create.sql && \
     satellite print-create-command >> /docker-entrypoint-initdb.d/create.sql
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -6,7 +6,7 @@ authors = [
 ]
 description = "Create, update and delete rows in a synthetic EMAP star schema"
 readme = "README.md"
-requires-python = ">=3.9"
+requires-python = ">=3.10"
 license = {file = "LICENSE"}
 classifiers = [
     "Programming Language :: Python :: 3"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -17,7 +17,7 @@ dependencies = [
     "coloredlogs==15.0.*",
     "networkx==2.8.*",
     "click==8.1.*",
-    "black==22.12.*",
+    "black==26.1.*",
     "psycopg2-binary==2.9.*",
     "pytest==7.2.*"
 ]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,6 @@
 [project]
 name = "satellite"
-version = "0.1.2"
+version = "0.2.0"
 authors = [
   { name="Tom Young" },
 ]

--- a/satellite/_column.py
+++ b/satellite/_column.py
@@ -17,7 +17,6 @@ from dataclasses import dataclass
 from satellite._log import logger
 from satellite._fake import fake
 
-
 if TYPE_CHECKING:
     from satellite._tables import Table
 

--- a/satellite/_column.py
+++ b/satellite/_column.py
@@ -11,7 +11,7 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Optional, Callable, TYPE_CHECKING
+from typing import Optional, Callable, Any, TYPE_CHECKING
 from dataclasses import dataclass
 
 from satellite._log import logger
@@ -54,6 +54,7 @@ class Column:
             "double": "real",
             "localdate": "date",
             "byte[]": "bytea",
+            "double[]": "real[]",
         }
 
         if self.java_type.lower() in java_to_sql_type_map:
@@ -65,16 +66,17 @@ class Column:
             )
             return "text"
 
-    @property
-    def format_specifier(self) -> str:
+    def sql_marshal(self, value: Any) -> str:
         if self.sql_type == "text":
-            return "'%s'"
+            return f"'{value}'"
         elif self.sql_type == "timestamptz" or self.sql_type == "date":
-            return "timestamp '%s'"
+            return f"timestamp '{value}'"
         elif self.sql_type == "bytea":
-            return "E'%s'"
+            return f"E'{value}'"
+        elif self.sql_type == "real[]":
+            return f"ARRAY {str(value)}"
 
-        return "%s"
+        return str(value)
 
     @property
     def is_foreign_key(self) -> bool:
@@ -114,5 +116,5 @@ class Column:
             return getattr(fake, self.sql_type)
 
         else:
-            logger.error(f"Have no provider for {self.sql_type}")
+            logger.error(f"Have no provider for {self.sql_type} in column {self.name}")
             return fake.default

--- a/satellite/_fake.py
+++ b/satellite/_fake.py
@@ -128,6 +128,9 @@ class _StarBaseProvider(BaseProvider):
     def standardised_vocabulary() -> None:
         return None
 
+    def values_array(self) -> list[float]:
+        return [self.real() for _ in range(10)]
+
 
 class _StarPersonProvider(PersonProvider, _StarBaseProvider):
     def firstname(self) -> str:

--- a/satellite/_schema.py
+++ b/satellite/_schema.py
@@ -89,14 +89,14 @@ class DatabaseSchema:
 
     def empty_table_create_command_for(self, table: Table) -> str:
         """Create a table for a set of data. Drop it if it exists"""
-
-        columns_name_and_type = ", ".join(
-            [col.definition_in_schema(self._name) for col in table.non_pk_columns]
-        )
+        non_pk_columns = ""
+        if len(table.non_pk_columns) > 0:
+            non_pk_columns += ", " + ", ".join(
+                [col.definition_in_schema(self._name) for col in table.non_pk_columns]
+            )
         return (
-            f"CREATE TABLE {self.schema_name}.{table.name} "
-            f"({table.primary_key_name} serial PRIMARY KEY, "
-            f"{columns_name_and_type});"
+            f"CREATE TABLE {self.schema_name}.{table.name}"
+            f"({table.primary_key_name} serial PRIMARY KEY{non_pk_columns});"
         )
 
     @staticmethod
@@ -118,7 +118,7 @@ class DatabaseSchema:
 
         for i in range(table.n_rows):
             values = ",".join(
-                column.format_specifier % self._decode_if_bytes(table[column][i])
+                column.sql_marshal(self._decode_if_bytes(table[column][i]))
                 if table[column][i] is not None
                 else "null"
                 for column in table.non_pk_columns

--- a/satellite/_schema.py
+++ b/satellite/_schema.py
@@ -118,9 +118,11 @@ class DatabaseSchema:
 
         for i in range(table.n_rows):
             values = ",".join(
-                column.sql_marshal(self._decode_if_bytes(table[column][i]))
-                if table[column][i] is not None
-                else "null"
+                (
+                    column.sql_marshal(self._decode_if_bytes(table[column][i]))
+                    if table[column][i] is not None
+                    else "null"
+                )
                 for column in table.non_pk_columns
             )
             string += f"  ({values}),\n"

--- a/satellite/_settings.py
+++ b/satellite/_settings.py
@@ -15,7 +15,6 @@ import os
 
 from typing import Any
 
-
 _default_values = {
     "STAR_SCHEMA_NAME": "star",
     "FAKER_SEED": "0",

--- a/satellite/_tables.py
+++ b/satellite/_tables.py
@@ -108,7 +108,7 @@ class Row(_TableChunk):
         return self[self.pk_column]
 
     @id.setter
-    def id(self, value: int):
+    def id(self, value: Optional[int]):
         self[self.pk_column] = value
 
     @property

--- a/satellite/_tables.py
+++ b/satellite/_tables.py
@@ -206,7 +206,6 @@ class Table(_TableChunk):
                 java_type=java_type,
                 parent_table_name=self.name,
             )
-
             self._data[column] = []
 
         logger.info(f"Created {self}")
@@ -264,7 +263,12 @@ class Tables(list):
     @classmethod
     def from_repo(cls, repo_url: str, branch_name: str) -> "Tables":
         """Create a list of tables by traversing files from a cloned git repo"""
-        excluded_suffixes = ["Core.java", "info.java", "TemporalFrom.java"]
+        excluded_suffixes = [
+            "Core.java",
+            "info.java",
+            "TemporalFrom.java",
+            "WaveformArray.java",
+        ]
         repo_path = Path("star_repo")
 
         if not repo_path.exists():

--- a/satellite/tests/test_table.py
+++ b/satellite/tests/test_table.py
@@ -16,7 +16,6 @@ import tempfile
 from pathlib import Path
 from satellite._tables import Table
 
-
 MINIMAL_TABLE_JAVA_FILE_LINES = (
     "@Entity",
     "@Table",

